design/operator_design.md

```diff
@@ -0,0 +1,129 @@
+# YugabyteDB Operator Design
+
+A cluster of YugaByte DB should be created using the `ybcluster.yugabytedb.yugabyte.com` custom resource definition. Below is a sample custom resource spec for creating a 3-master, 3 t-servers cluster using the CRD. The sample is followed by an explanation of different configuration options available on the YugaByte DB CRD.
+
+
+```yaml
+apiVersion: yugabytedb.yugabyte.com/v1alpha1
+kind: YBCluster
+metadata:
+  name: yugabytedb
+  namespace: yugabytedb
+spec:
+  image:
+    repository: yugabytedb/yugabyte     # Default.
+    tag: master                         # Default.
```

Comment:
[just be careful](https://helm.sh/docs/chart_best_practices/#images) about using `master` or [`latest` tags](https://kubernetes.io/docs/concepts/configuration/overview/#container-images); using `latest` is just a matter of time before it breaks someone who didn't pin it.

Reply:
Agreed. Though, the idea is user will override it, as needed. Pinning it to, probably current latest pinned tag (i.e. 1.3.2.0-b19), would mean user will have to override it always, to get the latest features.

Reply:
Right, but in general it's preferable to have to override the tag rather than take `latest`:

- if YB releases a `latest` which involves changes which users need to be aware of or new ways of how the operator has to be configured, then they're broken. That's a much worse experience than having users update the `tag` manually. In terms of user experience, stability is going to be significantly more valuable than features.
- Since the default `pullPolicy` is `IfNotPresent`, it means the "true" `latest` tag won't represent what YB publishes as `latest`. Since `latest` is just a tag, if the kubernetes hosts' docker daemon sees "`yugabytedb/yugabyte:latest`" cached, then it won't try to reach out to the upstream source to retrieve the "true" `latest`. It'll see whatever it has locally as `latest` and just use that.

That can get you into a weird spot where `latest` on one k8s host is actually `1.3.0-b17`, and then `latest` will represent `1.3.0.0-b17` until the docker image is purged from the hosts' docker image cache. Then let's say YB publishes `1.3.2.0-b19` and marks it as `latest`. In a given k8s cluster, if a _different_ k8s host comes online which doesn't have `yugabytedb/yugabyte:latest` cached, it'll pull it from the upstream repo and consider it "`latest`", e.g. to _that_ host in the k8s cluster, "`1.3.2.0-b19` is `latest`, but to the _other_ host, `1.3.0.0-b17` is `latest`... Then even within the same k8s cluster, there's uncertainty about which image is being used, because `latest` is relative.

see also: https://vsupalov.com/docker-latest-tag/

I presume this is something that would be configurable about the operator, but just something to consider. You might consider having the default be opting-in to using `master` rather than set it as default out of the box, just something to consider 👍

Reply:
Good point. Let's make `1.3.2.0-b19` as default.

Reply:
Done.